bpf/lib/overloadable_xdp.h

```diff
@@ -239,6 +239,32 @@ ctx_set_encap_info(struct xdp_md *ctx, __u32 src_ip, __be16 src_port,
 
 	return CTX_ACT_REDIRECT;
 }
+
+static __always_inline __maybe_unused int
+ctx_set_tunnel_opt(struct xdp_md *ctx, void *opt, __u32 opt_len)
+{
+	const __u32 geneve_off = ETH_HLEN + sizeof(struct iphdr) + sizeof(struct udphdr);
+	struct genevehdr geneve;
+
+	/* add free space after GENEVE header: */
+	if (ctx_adjust_hroom(ctx, opt_len, BPF_ADJ_ROOM_MAC, ctx_adjust_hroom_flags()) < 0)
```

Comment:
Noob question: this doesn't take `geneve_off` so how does the function know where to add space?

Reply:
See the corresponding change in `xdp.h`'s `ctx_adjust_hroom()`. It's completely custom and fragile - we just "know" how to behave for specific `hroom` values.